tensorflow/core/kernels/generate_box_proposals_op.cu.cc

```diff
@@ -0,0 +1,567 @@
+  return Status::OK();
+}
+
+// Allocate workspace for NMS operation
+Status AllocatePreNMSTempTensors(
+    OpKernelContext* context, Tensor* dev_image_prenms_boxes,
+    Tensor* dev_image_prenms_scores, Tensor* dev_image_boxes_keep_list,
+    Tensor* dev_postnms_rois, Tensor* dev_postnms_rois_probs,
+    Tensor* dev_prenms_nboxes, int num_images, int num_boxes_to_generate,
+    int box_dim, int post_nms_topn, int pre_nms_topn) {
+  auto d = context->eigen_gpu_device();
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({box_dim * num_boxes_to_generate}),
+      dev_image_prenms_boxes));
+  ResetTensor<float>(dev_image_prenms_boxes, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({num_boxes_to_generate}),
+      dev_image_prenms_scores));
+  ResetTensor<float>(dev_image_prenms_scores, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_INT32, TensorShape({num_boxes_to_generate}),
+      dev_image_boxes_keep_list));
+  ResetTensor<int32>(dev_image_boxes_keep_list, d);
+
+  const int max_postnms_nboxes = std::min(num_boxes_to_generate, post_nms_topn);
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT,
+      TensorShape({box_dim * num_images * max_postnms_nboxes}),
+      dev_postnms_rois));
+  ResetTensor<float>(dev_postnms_rois, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({num_images * max_postnms_nboxes}),
+      dev_postnms_rois_probs));
+  ResetTensor<float>(dev_postnms_rois_probs, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_INT32, TensorShape({num_images}), dev_prenms_nboxes));
+  ResetTensor<int32>(dev_prenms_nboxes, d);
+
+  return Status::OK();
+}
+
+// Initialize index and offset arrays.
+// num_images is the batch size.
+__global__ void InitializeDataKernel(const Cuda2DLaunchConfig config,
+                                     int* d_image_offsets,
+                                     int* d_boxes_keys_iota) {
+  const int image_size = config.virtual_thread_count.x;
+  const int num_images = config.virtual_thread_count.y;
+  CUDA_AXIS_KERNEL_LOOP(img_idx, config.virtual_thread_count.y, Y) {
+    CUDA_AXIS_KERNEL_LOOP(box_idx, config.virtual_thread_count.x, X) {
+      d_boxes_keys_iota[img_idx * image_size + box_idx] = box_idx;
+
+      // One 1D line sets the 1D data
+      if (box_idx == 0) {
+        d_image_offsets[img_idx] = image_size * img_idx;
+        // One thread sets the last+1 offset
+        if (img_idx == 0) d_image_offsets[num_images] = image_size * num_images;
+      }
+    }
+  }
+}
+
+}  // namespace
+
+class GenerateBoundingBoxProposals : public tensorflow::OpKernel {
+ public:
+  explicit GenerateBoundingBoxProposals(
+      tensorflow::OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("post_nms_topn", &post_nms_topn_));
+    CHECK_GT(post_nms_topn_, 0);
```

Comment:
@samikama remove all CHECKs and return proper error instead. It's not allowed and breaking the merge tests.

Reply:
@aaroey Is this a new requirement? I am seeing still there are a lot of references to CHECKs in TF code base. Can you be a bit more specific on which tests it is breaking?

Reply:
This is not new. It's required that new code (since >1 years ago I believed) should avoid checks. The merge process is blocked because of this check.

Reply:
That is strange since I can see commits as recent as last month that has same CHECK_* ops in the trunk. Are you sure that it is blocking the merge? If so how are some of the links below passed these tests? Pretty much all of them created after this pr is opened. Perhaps there should be some PR requirements document that everybody should adhere since if we use style/code similar to what is in TF or recent PRs, they are being marked as blockers. Could either somebody fix these in the TF so that other contributors can't use them as reference or can there be some requirements document created that everybody follows that document and prevent confusion and ease up non-google contributors life and improve their enthusiasm.

Thanks,
 
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/rocm/rocm_gpu_executor.cc#L421
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/host/host_gpu_executor.cc#L45
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/gpu/redzone_allocator.cc#L275
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/lib/core/py_func.cc#L193
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/eager/pywrap_tensor.cc#L121